spec/models/pet_spec.rb

```diff
@@ -54,31 +55,109 @@
-    expect(versions.second.reify).to be_a(Cat) # Sylvester
-    expect(versions.third.reify).to be_a(Cat) # Cheshire
-    expect(versions.fourth.reify).to be_a(Cat) # Cheshire that was destroyed
-
-    # Creating an object from the base class is correctly identified as "Animal"
-    expect(versions[5].reify).to be_an(Animal) # Muppets Drummer
-    expect(versions[6].reify).to be_an(Animal) # Animal that was destroyed
+  context "Older version entry present where item_type refers to the base_class" do
+    let(:cat) { Cat.create(name: "Garfield") }   # Index 0
+    let(:animal) { Animal.create }               # Index 4
+    let(:migrator) { ::PaperTrailSpecMigrator.new }
+
+    before do
+      # This line runs the `let` for :cat, creating two entries
+      cat.update_attributes(name: "Sylvester")   # Index 1 - second
+      cat.update_attributes(name: "Cheshire")    # Index 2 - third
+      cat.destroy                                # Index 3 - fourth
+      # With PT <= v9.2 a subclassed version's item_type referred to the base_class, but
+      # now it refers to the class itself.  In order to simulate an entry having been made
+      # in the old way, set one of our versions to be "Animal" instead of "Cat".
+      versions = PaperTrail::Version.order(:id)
+      versions.second.update(item_type: cat.class.base_class.name)
+
+      # This line runs the `let` for :animal, creating two entries
+      animal.update(name: "Muppets Drummer")    # Index 5
+      animal.destroy                            # Index 6
+    end
+
+    it "can reify a subclassed item" do
+      versions = PaperTrail::Version.order(:id)
+
+      # Still the reification process correctly brings back Cat since `species` is
+      # properly set to this sub-classed name.
+      expect(versions.second.reify).to be_a(Cat) # Sylvester
+      expect(versions.third.reify).to be_a(Cat) # Cheshire
+      expect(versions.fourth.reify).to be_a(Cat) # Cheshire that was destroyed
+
+      # Creating an object from the base class is correctly identified as "Animal"
+      expect(versions[5].reify).to be_an(Animal) # Muppets Drummer
+      expect(versions[6].reify).to be_an(Animal) # Animal that was destroyed
+    end
+
+    it "has a generator that builds migrations to upgrade older entries" do
+      # When using the has_many :versions association, it only finds versions in which
+      # item_type refers directly to the subclass name.
+      expect(cat.versions.count).to eq(3)
+      # To have has_many :versions work properly, you can generate and run a migration
+      # that examines all existing models to identify use of STI, then updates all older
+      # <= v9.2 version entries so they refer to the proper subclass.
+      # (This is the same as running:  rails g paper_trail:update_sti; rails db:migrate)
+      expect do
+        migrator.generate_and_migrate("paper_trail:update_sti")
+      end.to output(/Associated 1 record to Cat/).to_stdout
```

Comment:
I appreciate you testing the migration. Currently it is failing CI. I tried re-running the [job](https://travis-ci.org/paper-trail-gem/paper_trail/jobs/405184690) a few times, but it failed 4/5 runs. Don't worry if you can't get the expectation re: output passing, it's the data that matters :)

Reply:
Even without this output matcher in place, the association still hasn't properly taken place, so the later expect() will fail.  Have tracked this down to most likely being multiple Travis processes interfering with files in db/migrate (a) as processes are starting up and (b) when these generator tests are run.  Have created an advisory locking approach to deal with this concurrency which waits for up to 5 minutes for other processes to finish their things, and then (somewhat impatiently since perhaps another process has hung) jumps in and just takes an exclusive lock.

Reply:
>  Have created an advisory locking approach to deal with this concurrency which waits for up to 5 minutes for other processes to finish their things ..

Oh dear .. this complexity may outweigh the benefit of testing the migration.

Perhaps the meat of the migration could be moved into a plain ruby class that is easier to test?

Reply:
We create applications in which the integrity and security of customer data is mission-critical.  A proven test methodology is essential for this, and I will pay whatever price it takes in order to confirm a reliable system.

The locking strategy now in place here may allow generated migrations to always be reliably tested.  That's the goal anyway.  And for me it's not just for use here in PaperTrail with Travis, but also with our larger-scale projects that run on other CI tools with MANY more cores concurrently.